ilastik/workflows/neuralNetwork/_localLauncher.py

```diff
@@ -0,0 +1,54 @@
+import logging
+import sys
+import json
+import time
+import os
+import subprocess
+import tempfile
+
+logger = logging.getLogger(__name__)
+
+def wait(done, interval=0.1, max_wait=1):
+    start = time.time()
+
+    while True:
+        if done():
+            break
+
+        passed = time.time() - start
+
+        if passed > max_wait:
+            raise RuntimeError("timeout")
+        else:
+            time.sleep(interval)
+
+class LocalServerLauncher:
+    _executable_path: str
+    _process: subprocess.Popen
+
+    def __init__(self, executable_path: str) -> None:
+        self._executable_path = executable_path
+        self._process = None
+
+    def start(self):
+        if self._process:
```

Comment:
Any reason not to start the server in `__init__`? It would make it impossible for the server to be `.start`ed multiple times and it would remove the need for this check. It also makes it so that it's always legal to call `.stop`. And as a final bonus, it makes `_process` be correctly typed as `subprocess.Popen` instead of `Optional[subprocess.Popen]`

It's a shame one can't prevent a server from being `.stop`ped multiple times without ownership semantics =/

Reply:
This class can be used as a contextmanager, so start/stop could turn into `__enter__` and `__exit__`.

Reply:
> This class can be used as a contextmanager, so start/stop could turn into __enter__ and __exit__.

This class can't be used as context manager as it doesn't have clear scope located withing one function.

> Any reason not to start the server in __init__?

Starting the separate process in constructor seems like a unexpected behavior. Explicit verb in this case communicates intent better imo

Reply:
> This class can't be used as context manager as it doesn't have clear scope located withing one function.

Why not? It has only init, start and stop. Seems reasonable to acquire/release resources with contextmanager, or at least define a separate function that does it.

Reply:
>  Why not? It has only init, start and stop. Seems reasonable to acquire/release resources with contextmanager, or at least define a separate function that does it.

lifecycle of this class connected to a workflow class, so stop should be called from `cleanUp` method